src/client/tile.cpp

```diff
@@ -1824,6 +1912,109 @@ bool TextureSource::generateImagePart(std::string part_of_name,
 			}
 			pngimg->drop();
 		}
+		/*
+			[hsl:hue:saturation:lightness
+			or 
+			[colorizehsl:hue:saturation:lightness
+
+			Adjust the hue, saturation, and lightness of the base image. Like
+			"Hue-Saturation" in GIMP, but with 0 as the mid-point.
+			If colorize is true then all pixels will be converted to the
+			specified hue, but retain their saturation and lightness, like
+			"Colorize" in GIMP.
+			Hue should be from -180 to +180
+			Saturation and lightness are optional, and both from -100 to +100
+		*/
+		else if (str_starts_with(part_of_name, "[hsl:") || 
+		         str_starts_with(part_of_name, "[colorizehsl:")) {
+
+			if (baseimg == NULL) {
+				errorstream << "generateImagePart(): baseimg != NULL "
+					<< "for part_of_name=\"" << part_of_name
+					<< "\", cancelling." << std::endl;
+				return false;
+			}
+
+			Strfnd sf(part_of_name);
+			sf.next(":");
+			u32 hue = mystoi(sf.next(":"), -180, 180);
+			u32 saturation = sf.at_end() ? 0 : mystoi(sf.next(":"), -100, 100);
+			u32 lightness  = sf.at_end() ? 0 : mystoi(sf.next(":"), -100, 100);
```

Comment:
-100 does not fit into an u32, did you test this case?

Reply:
The [textureModifierTest mod](https://github.com/Treer/texturemodifiertest) tests all of these texture modifiers.

![negative-hsl](https://user-images.githubusercontent.com/6390507/179347571-58a58bd2-37ac-4e64-bc80-436526524d53.png)

I guess C++ and two's complement just don't care whether ints are signed or not :)
It's fixed now, thanks.

That test texture isn't great for testing saturation increases as it's already quite saturated, but I noticed raising sauturation turned the grey box in the centre red, which seemed odd, so I tried with the fantasy image.

It's doing what you expect for colourful parts of the image but the quality in grey parts of the image when increasing the saturation is so bad that I wonder if there's a bug in irrlicht's HSL implementation. The two images in the bottom left are made my Affinity, GIMP gives even nicer results but I think its scale just isn't going as stupidly high.

I think this will be a separate issue, but noting it here anyway

![hsl_saturation_irrlicht](https://user-images.githubusercontent.com/6390507/179348651-b1c8d196-31ad-4e46-87d9-8be5a725faa6.png)